src/registry/AvatarRegistry.sol

```diff
@@ -0,0 +1,631 @@
+        }
+
+        gasWei = uint256(feedValue);
+
+        /// @dev check latest oracle rate link/eth
+        (, feedValue, , timestamp, ) = LINK_ETH_FEED.latestRoundData();
+
+        if (block.timestamp - timestamp > CL_FEED_HEARTBEAT_LINK) {
+            revert StalePriceFeed(
+                address(LINK_ETH_FEED),
+                block.timestamp,
+                timestamp,
+                CL_FEED_HEARTBEAT_LINK
+            );
+        }
+
+        linkEth = uint256(feedValue);
+    }
+
+    /// @dev converts a gas limit value into link expressed amount
+    /// @param gasLimit amount of gas to provide the target contract when
+    /// performing upkeep
+    /// @return linkAmount amount of LINK needed to cover the job
+    function _getLinkAmount(uint256 gasLimit)
+        internal
+        view
+        returns (uint256 linkAmount)
+    {
+        (, IKeeperRegistry.Config memory _c, ) = CL_REGISTRY.getState();
+        (uint256 fastGasWei, uint256 linkEth) = _getFeedData();
+
+        uint256 adjustedGas = fastGasWei * _c.gasCeilingMultiplier;
+        uint256 weiForGas = adjustedGas * (gasLimit + REGISTRY_GAS_OVERHEAD);
+        uint256 premium = PPB_BASE + _c.paymentPremiumPPB;
+
+        /// @dev amount of LINK to carry one `performUpKeep` operation
+        // See: _calculatePaymentAmount
+        // https://etherscan.io/address/0x02777053d6764996e594c3E88AF1D58D5363a2e6#code#F1#L776
+        linkAmount =
+            // From Wei to Eth * Premium / Ratio
+            ((weiForGas * (1e9) * (premium)) / (linkEth)) +
+            (uint256(_c.flatFeeMicroLink) * (1e12));
+    }
+
+    /// @dev registers target avatar into the CL registry and saves
+    /// `upKeepId` into the mapping
+    /// @notice only callable from `performUpKeep` method via keepers
+    /// @param avatar contract avatar target to register
+    function _registerAndRecordId(address avatar) internal {
+        avatarsInfo[avatar].upKeepId = _registerUpKeep(
+            avatar,
+            avatarsInfo[avatar].gasLimit,
+            avatarsInfo[avatar].name
+        );
+    }
+
+    /// @dev checks if an avatar upKeepId is under-funded, helper in `checkUpKeep`
+    /// and `performUpKeep` methods
+    /// @param avatar contract address to verify if under-funded
+    function _isAvatarUpKeepUnderFunded(address avatar)
+        internal
+        view
+        returns (
+            uint256 upKeepId,
+            uint96 minUpKeepBal,
+            bool underFunded
+        )
+    {
+        if (avatar == address(this)) {
+            upKeepId = avatarMonitoringUpKeepId;
+        } else {
+            upKeepId = avatarsInfo[avatar].upKeepId;
+        }
+
+        /// @dev check onchain the min and current amounts to consider top-up
+        minUpKeepBal = CL_REGISTRY.getMinBalanceForUpkeep(upKeepId);
+        (, , , uint96 currentUpKeepBal, , , , ) = CL_REGISTRY.getUpkeep(
+            upKeepId
+        );
+
+        if (currentUpKeepBal <= minUpKeepBal * 3) {
```

Comment:
No magic numbers. Move `3` to a constant

Reply:
introduced a constant to described better its intent in c9847dd6a547fcb4bb2b63dad74f1bffb97174a5

```
uint256 internal constant MIN_ROUNDS_TOP_UP = 3;
```